### PR TITLE
Compile using heap memory

### LINKED ISSUE #32: Use stdin/stdout during compilation
When running a program, we currently rely on a temporary directory to store intermediate files. 
These IO operations during compilation are expensive and unnecessary. 
We should instead store the contents of generated files on the heap and pass them to the corresponding backends via stdin.

### PR DESCRIPTION
**Pull request by garritfra**
Fixes #32

### ToDo

- [x] Proposed feature/fix is sufficiently tested
- [x] Proposed feature/fix is sufficiently documented
- [x] The "Unreleased" section in the changelog has been updated, if applicable

### CODE REVIEW

File: src/command/run.rs [hunk 1]
```diff
@@ -36,11 +36,11 @@ pub fn run(target: Target, in_file: PathBuf) -> Result<(), String> {
                 Ok(process) => process,
             };
 
-            &process
+            process
                 .stdin
                 .unwrap()
                 .write_all(&buf)
-                .map(|_| "Could not write to nodejs process".to_string());
+                .expect("Could not write to nodejs process");
```
Comment (YerinAlexey) [inline src/command/run.rs:43]: Maybe write a message to stderr followed by `process::exit(1)` instead of panicing?

Comment (garritfra) [reply src/command/run.rs:43]: Good point👍

File: src/command/run.rs [hunk 2]
```diff
@@ -26,31 +26,28 @@ pub fn run(target: Target, in_file: PathBuf) -> Result<(), String> {
...
-                Ok(process) => process,
-            };
+                .map_err(|e| format!("Could not spawn Node.js process: {}", e.to_string()))?;
 
             process
                 .stdin
                 .unwrap()
                 .write_all(&buf)
-                .expect("Could not write to nodejs process");
+                .map_err(|e| format!("Could not write to Node.js process: {}", e.to_string()))?;
```
Comment (YerinAlexey) [inline src/command/run.rs:40]:
```suggestion
                .map_err(|e| {
                    format!("Could not write to Node.js process: {}", e);
                    std::process::exit(1);
                })?;
```

Comment (garritfra) [reply src/command/run.rs:40]: What would be the benefit of that? The error will be propagated upwards through the result, which will already cause the program to fail with a nice error message.

Comment (YerinAlexey) [reply src/command/run.rs:40]: Status codes. If program didn't finish successfully, the message is not enough, you also need to return non-0 exit code (usually 1 or 2). Shell and scripts usually take advantage of that:

- My prompt turns red if previous command failed
- Makefile recipe fails if command exits with non-0 code
- CI will also use that

Comment (YerinAlexey) [reply src/command/run.rs:40]: Nevermind

**Comment by YerinAlexey** [comment]
*About deleting that comment:* I forgot that `sb` exits with 1 when `Result` is `Err`

File: src/command/build.rs [hunk 3]
```diff
@@ -17,16 +13,20 @@ use std::io::Read;
...
     let mut buf = Box::new(Vec::new());
-    File::open(out_file)
-        .expect("Could not open file for writing")
-        .read_to_end(&mut buf)
-        .expect("Could not read from file");
-    build_to_buffer(target, in_file, &mut buf)
+    build_to_buffer(target, in_file, &mut buf)?;
+
+    File::create(out_file)
+        .map_err(|e| format!("Could not create output file: {}", e.to_string()))?
```
Comment (YerinAlexey) [inline src/command/build.rs:27]: It's not necessary to use `.to_string()` here as `format!` accepts string slices

File: src/builder/mod.rs [hunk 4]
```diff
@@ -122,7 +122,7 @@ impl Builder {
     pub(crate) fn generate(
         &mut self,
         target: generator::Target,
-        out_file: PathBuf,
+        buffer: &mut std::boxed::Box<impl Write>,
```
Comment (YerinAlexey) [inline src/builder/mod.rs:125]: Why you can't just use `impl Write`?

**Review by YerinAlexey — APPROVED** [review]

File: src/builder/mod.rs (hunk 4 above)
Comment (YerinAlexey) [reply src/builder/mod.rs:125]: Indeed, it compiles fine without it

```suggestion
        buffer: &mut impl Write,
```

**Comment by YerinAlexey** [comment]
@garritfra Does that look good?

**Comment by garritfra** [comment]
@YerinAlexey LGTM

**Pull request merged**